### PR TITLE
Riedl

### PR DESCRIPTION
**Pull request by jriedl**

### CODE REVIEW

File: cpp/pipelines/transfer_880_translations.cc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:25]: Why the blank line, we never do that anywhere else.

File: cpp/pipelines/transfer_880_translations.cc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#include "MARC.h"
+#include "util.h"
+
+
+namespace {
+
+
+    const std::string UNDEFINED_LANGUAGE("Undefined");
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:33]: Why the indentation, we never do that anywhere else.

File: cpp/pipelines/transfer_880_translations.cc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    [[noreturn]] void Usage() {
+        ::Usage("authority_input autority_output");
+    }
+
+    std::string GetSubfieldAValueFromIxTheoTranslationField(const MARC::Record::Field &field) {
+        return field.getSubfields().getFirstSubfieldWithCode('a');
+    }
+
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:42]: Missing black line between functions.

File: cpp/pipelines/transfer_880_translations.cc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    std::string Get9ZValueFromIxTheoTranslationField(const MARC::Record &record, const MARC::Record::Field &field) {
+        const auto _9Z_values = field.getSubfields().extractSubfieldsAndNumericSubfields("9Z");
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:44]: Please use parentheses for variable initialisation.

File: cpp/pipelines/transfer_880_translations.cc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        if (_9Z_values.empty())
+            LOG_ERROR("No valid 9Z content for record " + record.getControlNumber());
+        return _9Z_values[0];
+    }
+
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:49]: Missing blank line.

File: cpp/pipelines/transfer_880_translations.cc [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    void ProcessRecords(MARC::Reader * const marc_reader, MARC::Writer * const marc_writer) {
+        unsigned count(0), modified(0);
+        std::vector<MARC::Subfields> new_750_entries;
+        while(MARC::Record record = marc_reader->read()) {
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:53]: Missing space.

File: cpp/pipelines/transfer_880_translations.cc [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            if (not new_750_entries.empty()) {
+               for (const auto &new_750_subfields : new_750_entries) {
+                   if (not record.insertFieldAtEnd("750", new_750_subfields, ' ', '7'))
+                       LOG_ERROR("Could not insert field for record " + record.getControlNumber());
+               }
+               ++modified;
+            }
+            marc_writer->write(record);
+        }
+        std::cout << "Modified " << modified << " record of " << count << " altogether " << '\n';
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:84]: Please use LOG_INFO.

File: cpp/pipelines/transfer_880_translations.cc [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            LOG_ERROR("No valid 9Z content for record " + record.getControlNumber());
+        return _9Z_values[0];
+    }
+
+    void ProcessRecords(MARC::Reader * const marc_reader, MARC::Writer * const marc_writer) {
+        unsigned count(0), modified(0);
+        std::vector<MARC::Subfields> new_750_entries;
+        while(MARC::Record record = marc_reader->read()) {
+            ++count;
+            if (not record.hasFieldWithTag("880"))
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:55]: Don't we lose these records in the output?

File: cpp/pipelines/transfer_880_translations.cc [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+
+} // unnamed namespace
+
+
+int Main(int argc, char **argv) {
+    if (argc != 3)
+        Usage();
+
+    std::unique_ptr<MARC::Reader> marc_reader(MARC::Reader::Factory(argv[1]));
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:95]: Can be const.

File: cpp/pipelines/transfer_880_translations.cc (hunk 8 above)
Comment (jriedl) [reply cpp/pipelines/transfer_880_translations.cc]: Thanks for the hint.

File: cpp/pipelines/transfer_880_translations.cc [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include "MARC.h"
+#include "util.h"
+
+
+namespace {
+
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:30]: A blank line is missing.

File: cpp/pipelines/transfer_880_translations.cc [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+const std::string UNDEFINED_LANGUAGE("Undefined");
+
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:32]: Ditto.  We always have exactly 2 blank lines between file-scope entities, declarations functions etc.  There are *no* special rules!

File: cpp/pipelines/transfer_880_translations.cc [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        for (const auto &field : record) {
+            if (field.getTag() == "880" and field.hasSubfieldWithValue('2', "IxTheo")) {
+                std::string language_code(UNDEFINED_LANGUAGE);
+                if (field.hasSubfieldWithValue('6', "750-01/Hant"))
+                    language_code = "hant";
+                else if (field.hasSubfieldWithValue('6', "750-01/Hans"))
+                    language_code = "hans";
+                if (language_code != UNDEFINED_LANGUAGE) {
+                    new_750_entries.emplace_back(MARC::Subfields({
+                        {'a', GetSubfieldAValueFromIxTheoTranslationField(field) },
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:70]: Missing space after the opening brace, here and below.

File: cpp/pipelines/transfer_880_translations.cc [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+} // unnamed namespace
+
+
+int Main(int argc, char **argv) {
+    if (argc != 3)
+        Usage();
+
+    const std::unique_ptr<MARC::Reader> marc_reader(MARC::Reader::Factory(argv[1]));
+    std::unique_ptr<MARC::Writer> marc_writer(MARC::Writer::Factory(argv[2]));
```
Comment (ruschein) [inline cpp/pipelines/transfer_880_translations.cc:99]: This can be const, too.

**Pull request merged**